tests/cxx/implicit_ctor-d1.h

```diff
@@ -21,15 +22,44 @@ int ImplicitCtorFn(IndirectWithImplicitCtor);
 // IWYU: IndirectWithImplicitCtor is...*implicit_ctor-i2.h.*for autocast
 int ImplicitCtorRefFn(const IndirectWithImplicitCtor&);
 
+// Autocast assurance for header-defined functions still makes sense as opposed
+// to function definitions in source files.
+// IWYU: IndirectWithImplicitCtor needs a declaration
+// IWYU: IndirectWithImplicitCtor is...*implicit_ctor-i2.h.*for autocast
+inline int InlineImplicitCtorRefFn(const IndirectWithImplicitCtor&) {
```

Comment:
Maybe, cover more test cases? E.g., when "autocast" assurance is inhibited by forward-declaring...

Reply:
I believe that's covered by `iwyu_stricter_than_cpp*`

Reply:
But there is no function definitions in `iwyu_stricter_than_cpp-autocast.h`, only declarations...

Reply:
I see. I think you could add that example in `iwyu_stricter_than_cpp-autocast.h` -- the forward-decls are already available there, so it should just be a matter of adding a small inline function.

Ah, or copy the `Fn` there and make it an inline definition, so it covers all the same cases.

Reply:
I've added new cases into `iwyu_stricter_than_cpp`. But I'm not sure that they aren't just a redundant garbage, which multiplies possible subsequent test case changes by three...  Please take a look.

Reply:
It does get pretty complicated with all the different types. Is it really necessary to use distinct types for every parameter?

Reply:
I think so, because various combinations of fwd-decl status and `#include` status should be tested. Or did you mean separate `DirectStruct*`s for every declaration? They must be different because requirement to `#include` them is only tested through the explanatory comment in "IWYU SUMMARY", because they are included directly. If it were just a single type, one could broke all reportings but one and still have the test passed.

Reply:
Aha. That seems like a weakness in the test design; wouldn't it be better to use one type, but not have it directly included? So that there were line diagnostics? I'm not suggesting you should do it, I'm just curious if the test is really its best self.

Reply:
No, the point is exactly in that the type definition is included directly. This is one of the two "type provision" rules (a type is considered as [should be] provided if it is not fwd-declared or is _directly_ included) which are tested in `iwyu_stricter_than_cpp`.

Reply:
Oh, right. I overlooked the context.

I'm almost inclined to reduce coverage to keep the tests less noisy here. I'll leave it to you to decide.